### PR TITLE
EIP-2645: Adding Hierachical Wallet For Layer-2

### PR DESCRIPTION
**Pull request by louisguthmannStarkWare**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

---
This change is [Reviewable](https://reviewable.io/reviews/ethereum/eips/2645)

### CODE REVIEW

File: EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+eip: <to be assigned>
```
Comment (axic) [inline EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md:2]: 2645.

Also please rename the file.

File: EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+title: Hierarchical Deterministic Wallet for Computation Integrity Proof (CIP) Layer-2
+author: Tom Brand (tom@starkware.co), Louis Guthmann (louis@starkware.co)
+discussions-to: https://ethereum-magicians.org/t/hierarchical-deterministic-wallet-for-computation-integrity-proof-cip-layer-2/4286
+status: Draft
+type: Standard Track 
+category (*only required for Standard Track): ERC
```
Comment (axic) [inline EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md:8]: Please remove the `(*..)` comment.

File: EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md (hunk 1 above)
Comment (guthlStarkware) [reply EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md]: Thanks for the comment!
Done

File: EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md (hunk 2 above)
Comment (guthlStarkware) [reply EIPS/Hierarchical Deterministic Wallet for CIP Layer-2.md]: Done

File: EIPS/EIP-2645 - Hierarchical Deterministic Wallet for CIP Layer-2.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. These curves are Baby Jubjub (in the context of the BN254 SNARK approach) and Arya in (the context of StarkEx). Extensive work has been done to provide secure ways to derive private keys. We leverage this work to define an efficient way to securely derive private keys from existing infrastructure, as well as creating domain separation between Layer-2 applications.
+
+## Abstract
+We provide a Derivation Path allowing a user to derive hierarchical keys for Layer-2 solutions depending on the zk-technology, the application, the user’s Layer-1 address, as well as an efficient grinding method to enforce the private key distribution within the curve domain.
+
+## Motivation
+TDB
+
+## Specification
+Starkware keys are derived with the following BIP43-compatible derivation path, with direct inspiration from BIP44:
```
Comment (axic) [inline EIPS/EIP-2645 - Hierarchical Deterministic Wallet for CIP Layer-2.md:22]:
```suggestion
Starkware keys are derived with the following [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)-compatible derivation path, with direct inspiration from BIP44:
```

File: EIPS/eip-2645.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
+---
+eip: 2645
+title: Hierarchical Deterministic Wallet for Computation Integrity Proof (CIP) Layer-2
+author: Tom Brand (tom@starkware.co), Louis Guthmann (louis@starkware.co)
+discussions-to: https://ethereum-magicians.org/t/hierarchical-deterministic-wallet-for-computation-integrity-proof-cip-layer-2/4286
+status: Draft
+type: Standard Track
```
Comment (axic) [inline EIPS/eip-2645.md:7]:
```suggestion
type: Standards Track 
```

File: EIPS/eip-2645.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+---
+
+## Simple Summary
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. These curves are Baby Jubjub (in the context of the BN254 SNARK approach) and Arya in (the context of StarkEx). Extensive work has been done to provide secure ways to derive private keys. We leverage this work to define an efficient way to securely derive private keys from existing infrastructure, as well as creating domain separation between Layer-2 applications.
+
+## Abstract
+We provide a Derivation Path allowing a user to derive hierarchical keys for Layer-2 solutions depending on the zk-technology, the application, the user’s Layer-1 address, as well as an efficient grinding method to enforce the private key distribution within the curve domain.
+
+## Specification
+Starkware keys are derived with the following [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)-compatible derivation path, with direct inspiration from BIP44:
```
Comment (axic) [inline EIPS/eip-2645.md:19]:
```suggestion
Starkware keys are derived with the following [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)-compatible derivation path, with direct inspiration from [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki):
```

File: EIPS/eip-2645.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+```
+m / purpose' / layer' / application' / eth_address_1' / eth_address_2' / index
+```
+where:
+* `m` - the seed.
+* `purpose` - the EIP number.
```
Comment (axic) [inline EIPS/eip-2645.md:25]:
```suggestion
* `purpose` - `2545` (the number of this EIP).
```

File: EIPS/eip-2645.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+This EIP specifies two aspects of keys derivation in the context of Hierarchical Wallets:
+- Derivation Path
+- Grinding Algorithm to enforce a uniform distribution over the elliptic curve.
+The derivation path is defined to allow efficient keys separation based on technology and application while maintaining a 1-1 relation with the Layer-1 wallet. In such a way, losing EIP-2645 wallets falls back to losing the Layer-1 wallet.
+
+## Backwards Compatibility
+This standard is backward compatible with BIP43.
+
+## Security Considerations
+As explained previously, EIP-2645 has been defined to maintain separation of keys while providing foolproof logic on key derivation.
```
Comment (axic) [inline EIPS/eip-2645.md:60]:
```suggestion
As explained previously, this EIP has been defined to maintain separation of keys while providing foolproof logic on key derivation.
```

File: EIPS/eip-2645.md [hunk 8]
```diff
@@ -15,18 +15,15 @@ In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK
 ## Abstract
 We provide a Derivation Path allowing a user to derive hierarchical keys for Layer-2 solutions depending on the zk-technology, the application, the user’s Layer-1 address, as well as an efficient grinding method to enforce the private key distribution within the curve domain.
 
-## Motivation
```
Comment (axic) [inline EIPS/eip-2645.md]: Please add back the motivation.

File: EIPS/eip-2645.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+## Motivation
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. Extensive work has been done to make it secure on Bitcoin via [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki), [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) and [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki). Those protosals have become the standard for wallets in the entire industry, independantly on the associated blockchain. As Layer-2 solutions are taking off, it was a necessary requirement to maintain the same standard and security in this new space. StarkWare, in the context of the launch of their first products, needed to explore the existing solutions and adapt them to match its requirements. While working with wallets, it felt that the effort needed to be shared to benefit the community and get additional feedback. This present EIP is the result of this work.
```
Comment (edsonayllon) [inline EIPS/eip-2645.md:19]: "independantly" should be "independently."

File: EIPS/eip-2645.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+## Specification
+Starkware keys are derived with the following [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)-compatible derivation path, with direct inspiration from [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki):
+```
+m / purpose' / layer' / application' / eth_address_1' / eth_address_2' / index
+```
+where:
+* `m` - the seed.
+* `purpose` - `2545` (the number of this EIP).
+* `layer` - the 31 lowest bits of sha256 on the layer name. Serve as a domain seperator between different technologies. In the context of `starkex`, the value would be `579218131`.
```
Comment (edsonayllon) [inline EIPS/eip-2645.md:29]: "seperator" should be "separator." It's misspelled twice in this list.

**Comment by guthlStarkware** [comment]
How can I reopen and get a merge on this?

**Comment by MicahZoltu** [comment]
I'll take a look at it later today to see if it is good to merge as draft.

File: EIPS/eip-2645.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
+---
+eip: 2645
+title: Hierarchical Deterministic Wallet for Computation Integrity Proof (CIP) Layer-2
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:3]: Can we shorten this title at all?  This feels quite a bit too long for a title and really would be better as a Simple Summary.

File: EIPS/eip-2645.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+author: Tom Brand (tom@starkware.co), Louis Guthmann (louis@starkware.co)
+discussions-to: https://ethereum-magicians.org/t/hierarchical-deterministic-wallet-for-computation-integrity-proof-cip-layer-2/4286
+status: Draft
+type: Standards Track 
+category: ERC
+created: 2020-05-13
+---
+
+## Simple Summary
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. These curves are Baby Jubjub (in the context of the BN254 SNARK approach) and Arya in (the context of StarkEx). Extensive work has been done to provide secure ways to derive private keys. We leverage this work to define an efficient way to securely derive private keys from existing infrastructure, as well as creating domain separation between Layer-2 applications.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:13]: Simple Summary should be of length similar to an email subject line or forum thread title.  This is currently far too long, and most of the content is better suited for motivation or abstract.

File: EIPS/eip-2645.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+## Abstract
+We provide a Derivation Path allowing a user to derive hierarchical keys for Layer-2 solutions depending on the zk-technology, the application, the user’s Layer-1 address, as well as an efficient grinding method to enforce the private key distribution within the curve domain.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:16]: Consider including the derivation path here.  Abstract should be a human readable and very terse version of the specification.  Ideally, someone can read the abstract and get the gist of the specification without having to read anything else.  Usually minor details and edge cases are ignored, and only the core of the specification is included in plain English.  This abstract is pretty solid as is, would just be nice to throw `m / purpose' / layer' / application' / eth_address_1' / eth_address_2' / index` in there somewhere since that is the core of the specification.

File: EIPS/eip-2645.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+## Specification
+Starkware keys are derived with the following [BIP43](https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki)-compatible derivation path, with direct inspiration from [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki):
+```
+m / purpose' / layer' / application' / eth_address_1' / eth_address_2' / index
+```
+where:
+* `m` - the seed.
+* `purpose` - `2645` (the number of this EIP).
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:28]: I believe BIP 43 recommends filing a BIP to reserve this number, rather than an EIP: 
> We encourage different schemes to apply for assigning a separate BIP number and use the same number for purpose field, so addresses won't be generated from overlapping BIP32 spaces.

File: EIPS/eip-2645.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+This EIP specifies two aspects of keys derivation in the context of Hierarchical Wallets:
+- Derivation Path
+- Grinding Algorithm to enforce a uniform distribution over the elliptic curve.
+The derivation path is defined to allow efficient keys separation based on technology and application while maintaining a 1-1 relation with the Layer-1 wallet. In such a way, losing EIP-2645 wallets falls back to losing the Layer-1 wallet.
+
+## Backwards Compatibility
+This standard is backward compatible with BIP43.
+
+## Security Considerations
+As explained previously, this EIP has been defined to maintain separation of keys while providing foolproof logic on key derivation.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:65]:
```suggestion
This EIP has been defined to maintain separation of keys while providing foolproof logic on key derivation.
```
Suggest keeping the specification tight and technical, rather than loose and conversational.

File: EIPS/eip-2645.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+Nota Bene: At each round, the probability for a key to be greater than (N - (N % n)) is < 2^(-5).
+
+## Rationale
+This EIP specifies two aspects of keys derivation in the context of Hierarchical Wallets:
+- Derivation Path
+- Grinding Algorithm to enforce a uniform distribution over the elliptic curve.
+The derivation path is defined to allow efficient keys separation based on technology and application while maintaining a 1-1 relation with the Layer-1 wallet. In such a way, losing EIP-2645 wallets falls back to losing the Layer-1 wallet.
+
+## Backwards Compatibility
+This standard is backward compatible with BIP43.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:62]: I don't think backward compatible is the right term here, this EIP complies with BIP43.

I do think you should probably mention that the current popular derivation paths for Ethereum are incompatible with this, and mention that wallets will potentially need to prompt the user for derivation path if they support this and other Ethereum derivation paths.

File: EIPS/eip-2645.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+## Motivation
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. Extensive work has been done to make it secure on Bitcoin via [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki), [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) and [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki). Those protosals have become the standard for wallets in the entire industry, independently on the associated blockchain. As Layer-2 solutions are taking off, it was a necessary requirement to maintain the same standard and security in this new space. StarkWare, in the context of the launch of their first products, needed to explore the existing solutions and adapt them to match its requirements. While working with wallets, it felt that the effort needed to be shared to benefit the community and get additional feedback. This present EIP is the result of this work.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:19]: > StarkWare, in the context of the launch of their first products, needed to explore the existing solutions and adapt them to match its requirements. While working with wallets, it felt that the effort needed to be shared to benefit the community and get additional feedback. This present EIP is the result of this work.

We have been having problems with people using EIPs as a means of advertising their products, so we generally prefer to not mention/reference external products in EIPs unless it is critical to the EIP.

Recommend removing these last few sentences, as they are more history than motivation.  Alternatively, consider rewording this to be less historic and more present-tense motivation.

In general, the motivation section should target an audience of someone 5-10 years from now looking at this EIP and wondering why they might want to implement it.  Listing potential use cases/applications are a great example of this.

**Review by MicahZoltu — APPROVED** [review]
I left some comments, but nothing stands out as blocking this from being merged as a draft.

Sorry it took so long for an editor to get to it, we have historically had problems with editors not reviewing EIPs but we are trying to do better!

File: EIPS/eip-2645.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
+---
+eip: 2645
+title: Hierarchical Deterministic Wallet for Computation Integrity Proof (CIP) Layer-2
+author: Tom Brand (tom@starkware.co), Louis Guthmann (louis@starkware.co)
+discussions-to: https://ethereum-magicians.org/t/hierarchical-deterministic-wallet-for-computation-integrity-proof-cip-layer-2/4286
+status: Draft
+type: Standards Track
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:7]:
```suggestion
type: Standards Track
```
CI is *very* picky about the formatting of the frontmatter, this change is necessary to merge this as draft.

**Review by DONALDSULLENDER — APPROVED** [review]

File: EIPS/eip-2645.md (hunk 11 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Reduced to Hierarchical Deterministic Wallet for Layer-2

File: EIPS/eip-2645.md (hunk 12 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Done

File: EIPS/eip-2645.md (hunk 13 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Done

File: EIPS/eip-2645.md (hunk 17 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Removed all mention of Starkware

File: EIPS/eip-2645.md (hunk 16 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Done

File: EIPS/eip-2645.md (hunk 15 above)
Comment (louisguthmannStarkWare) [reply EIPS/eip-2645.md]: Done

**Review by louisguthmannStarkWare — COMMENTED** [review]
@MicahZoltu Took your comments into consideration. BTY

File: EIPS/eip-2645.md [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+category: ERC
+created: 2020-05-13
+---
+
+## Simple Summary
+In the context of Computation Integrity Proof (CIP) Layer-2 solutions such as ZK-Rollups, users are required to sign messages on new elliptic curves optimized for those environnements. We leverage existing work on Key Derivation ([BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki), [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) and [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki)) to define an efficient way to securely produce CIP L2s private keys, as well as creating domain separation between Layer-2 applications.
```
Comment (MicahZoltu) [inline EIPS/eip-2645.md:13]: This is much too long for Simple Summary.  Think of what you would put in an email subject line, or a forum thread topic, or a GitHub PR title.

**Review by MicahZoltu — APPROVED** [review]
